// gp4/dist/core/response.js
import { RequestError } from './errors.js';
import { decode } from './decode/gp4.js';
export const isResponseOk = (response) => {
   const { statusCode } = response;
   const { followRedirect } = response.request.options;
   const shouldFollow = typeof followRedirect === 'function' ? followRedirect(response) : followRedirect;
   const limitStatusCode = shouldFollow ? 299 : 399;
   return (statusCode >= 200 && statusCode <= limitStatusCode) || statusCode === 304;
};
export class ParseError extends RequestError {
   constructor(error, response) {
      const { options } = response.request;
      super(`${error.message} in "${options.url.toString()}"`, error, response.request);
      this.name = 'ParseError';
      this.code = 'ERR_BODY_PARSE_FAILURE';
   }
}
export const parseBody = (response, responseType, parseJson, encoding) => {
    const { rawBody } = response;
    let body;
    body = decode(rawBody.toString(encoding)) ?? rawBody.toString(encoding);
    try {
        if (responseType === 'text') {
            return body;
        }
        else if (responseType === 'json') {
            return parseJson(body);
        }
        else if (responseType === 'buffer') {
            return rawBody;
        }
    }
    catch (e) {
        throw new ParseError(e, response);
    }
    throw new ParseError({
        message: `Unknown body type '${responseType}'`,
        name: 'Error',
    }, response);
};
